ReSharper.FSharp/src/FSharp.Psi.Services/src/Util/SignatureFileUtil.fs

```diff
@@ -0,0 +1,172 @@
+module JetBrains.ReSharper.Plugins.FSharp.Psi.Services.Util.SignatureFile
+
+open System
+open FSharp.Compiler.Symbols
+open JetBrains.ReSharper.Plugins.FSharp.Psi
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Util
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl
+
+let private (|DeclaredNameInPattern|_|) (pat:IFSharpPattern) =
+    match pat with
+    | :? ILocalReferencePat as pat -> Some pat.DeclaredName
+    | :? ITypedPat as tp ->
+        match tp.Pattern.IgnoreInnerParens() with
+        | :? ILocalReferencePat as pat -> Some pat.DeclaredName
+        | _ -> None
+    | _ -> None
+
+type SignatureBindingResponse =
+    {
+        SigDeclNode: IBindingSignature
+        OpenStatements: IOpenStatement list
+        SigModule: INamedModuleDeclaration
+    }
+    
+    member this.SigFile = this.SigModule.Parent :?> IFSharpSigFile
+
+let tryMkBindingSignature
+    (letBindings: ILetBindings)
+    (moduleDecl: IModuleDeclaration)
+    : SignatureBindingResponse option
+    =
+    if isNull moduleDecl then None else
+    
+    let allDecls = moduleDecl.DeclaredElement.GetDeclarations() |> Seq.cast<INamedModuleDeclaration> |> Seq.toArray
+    if allDecls.Length <> 2 then None else
+
+    let implDecl = Array.find (fun (nmd: INamedModuleDeclaration) ->
+        match nmd.Parent with
+        | :? IFSharpImplFile -> true
+        | _ -> false) allDecls
+
+    let sigDecl = Array.find (fun (nmd: INamedModuleDeclaration) ->
+        match nmd.Parent with
+        | :? IFSharpSigFile -> true
+        | _ -> false) allDecls
+
+    let elementFactory = sigDecl.CreateElementFactory()
+    let binding = letBindings.Bindings.First()
+    let refPat = binding.HeadPattern.As<IReferencePat>()
+    let name = refPat.ReferenceName.Identifier.Name
```

Comment:
There will likely be other features and cases where `tryMkBindingSignature` is going to be used, so we should not assume it's only called on `IReferencePat` pattern bindings. This will throw NRE.